### DISCUSSION TITLE
Non-initial characters include any decimal digit (0-9)

### DISCUSSION THREAD

**Discussion by mikeparisstuff** [kdl-org/kdl#158] [Q&A] [answered]
First of all, I think KDL looks awesome so great work. I was reading through the spec and believe I found a difference between KDL and SDLang that impacts one of the use cases mentioned on the SDLang homepage that I didn't see mentioned in the readme. Curious if I am correct and, if so, as to your reasoning.

On the SDLang homepage they mention being able to represent a matrix like:

```
matrix {
  1 0 0
  0 1 0
  0 0 1
}
```

However in KDL, given that the elements within a child block are nodes, and nodes must either begin with an identifier or a quoted string, wouldn't this not be allowed within KDL given the non-initial character constraints? According to the spec, I would be able to do:

```
matrix {
  "1"; "0"; "0";
  "0"; "1"; "0";
  "0"; "0"; "1";
}
```

but then given that I am representing a matrix that I would like to interpret as numbers, this changes the meaning. I also realize that I could do something like shown below but this could quickly become verbose.

```
matrix {
  value 1; value 0; value 0;
  value 0; value 1; value 0;
  value 0; value 0; value 1;
}

// or perhaps more succinctly
matrix {
  row 1 0 0
  row 0 1 0
  row 0 0 1
}
```

Curious if you think this is a valid use case and if it justifies allowing identifiers to be numbers, or, at least, to start with numbers. I could see a similar situation as to what you specify for strings in that a node can be `ident="val"` or `"ident"="val"` or `1="val"` or plain `1`. If this were the case, given the terminator rule it would still be a bit different, but it gets us closer:

```
matrix {
  1; 0; 0;
  0; 1; 0;
  0; 0; 1;
}
```

**Comment by zkat** [comment, marked as answer]
This was intentionally removed because I found anonymous nodes kinda awkward to work with, and then I added string-name nodes, which conflict with that feature.

The SDLang example basically translates to this, internally:

```
matrix {
  anon 1 0 0
  anon 0 1 0
  anon 0 0 1
}
```

My personal recommendation when you just want an "anonymous" node is to use `-` as a name, which seems to be something that's increasingly popular among the early adopter crowd:

```kdl
matrix {
  - 1 0 0
  - 0 1 0
  - 0 0 1
}
```

I actually prefer this because I find that the matrix example from SDLang obfuscates the fact that these are just 3 nodes and makes it seem like you can arbitrarily organize values like that? Does that make sense?

**Reply by larsgw** [reply]
Just wanted to add quickly that this is also better, in my opinion at least, than the example with semicolons and numeric identifiers because now you don't need to specify the dimensions of the matrix.

**Reply by mikeparisstuff** [reply]
Thanks for the responses, and yes this makes sense. The point on obfuscating that it is only 3 nodes is a good one. I like the `-` syntax for an anonymous node but given the overlap with yaml arrays perhaps someone will find it confusing. Then again, this isn't yaml so why one would think it shares the same rules, idk. Either way, it does make the point clear that you can easily represent said thing, with minimal overhead, and that retains the semantics of the numbers. Good enough for me.
